fullstack-cert/typescript-projects/fortune-app/src/script.ts

```diff
@@ -0,0 +1,265 @@
+      text: getElement(".text")!,
+    };
+    // 2) Fetch data
+    this.fetchCardsData();
+    // 3) Bind events
+    //    (kept separate for readability and easier testing)
+    this.initializeEventListeners();
+  }
+
+  /**
+   * Fetch and store tarot card data from the remote JSON.
+   * Logs an error to the console if the request fails.
+   */
+  private async fetchCardsData() {
+    try {
+      const response = await fetch(`${CDN_URL}/card_data.json`);
+      const data: Deck = await response.json();
+      this.cards = data.cards;
+    } catch (error) {
+      console.error("Error fetching cards:", error);
+    }
+  }
+
+  /** Bind all UI event listeners used by the app */
+  private initializeEventListeners(): void {
+    // Event listeners
+    this.elements.singleCardBtn.addEventListener("click", () =>
+      this.singleCardSelected(),
+    );
+    this.elements.multipleCardsBtn.addEventListener("click", () =>
+      this.multipleCardSelected(),
+    );
+    this.elements.newReadingBtn.addEventListener("click", () =>
+      this.newReading(),
+    );
+    document.addEventListener("click", (e: Event) => this.showFortune(e));
+  }
+
+  /**
+   * Show a single randomly selected card.
+   * Hides the initial buttons and reveals the single-card view.
+   */
+  singleCardSelected() {
+    hideElements(
+      this.elements.singleCardBtn,
+      this.elements.multipleCardsBtn,
+      this.elements.multipleCard,
+      this.elements.text,
+      this.elements.headerTitle,
+    );
+
+    const isReversed = Math.random() < 0.5;
+    const chosenCard = getRandomItem(this.cards);
+
+    this.elements.singleCard.innerHTML = renderCard(
+      "Your card",
+      isReversed,
+      chosenCard.name_short,
+      chosenCard.img,
+    );
+    this.elements.multipleCard.innerHTML = "";
+
+    showElements(this.elements.singleCard, this.elements.fortuneContainer);
+  }
+
+  /**
+   * Show three cards (past, present, future).
+   * Hides the single-card elements and reveals the multi-card view.
+   */
+  multipleCardSelected() {
+    hideElements(
+      this.elements.singleCard,
+      this.elements.singleCardBtn,
+      this.elements.multipleCardsBtn,
+      this.elements.headerTitle,
+    );
+    showElements(
+      this.elements.multipleCard,
+      this.elements.fortuneContainer,
+      this.elements.text,
+    );
+
+    this.elements.multipleCard.innerHTML = Object.values(DrawingType)
+      .map((type) => {
+        const isReversed = Math.random() < 0.5;
+        const card = getRandomItem(this.cards);
+        return renderCard(type, isReversed, card.name_short, card.img);
+      })
+      .join("");
+  }
+
+  /**
+   * Populate the fortune description panel when a card is clicked.
+   * Uses the card's data-id attribute to locate the matching card data.
+   */
+  showFortune(e: Event) {
+    const target = e.target as HTMLElement | null;
+    const cardElement = target?.closest(
+      ".card_container",
+    ) as HTMLElement | null;
```

Comment:
We should typeguard here, too, rather than casting.

Reply:
> question: what is the reason some cards appear capsized (and turns face-up when hovering)?

Based on my understanding, Tarot can be revealed face up and face down with different meanings, and the flip on hover is just a quality of life feature to make it easier to identify on mouse-over.

Reply:
Exactly.